### ISSUE TITLE
streamState feature request

### ISSUE THREAD

**Issue by skynet1** [AccelerateHS/accelerate-llvm#37]
I am submitting a...
- [ ] bug report
- [x] feature request
- [ ] support request => you might also like to ask your question on the [mailing list](mailto:accelerate-haskell@googlegroups.com) or [gitter chat](https://gitter.im/AccelerateHS/Lobby).

## Description

I'm trying to write a function that can do the same operation over and over again on a 
huge list of arrays. The list is too big to fit on the GPU, and I want to send the arrays one at a time
to the GPU, and it can modify it's internal state each time.

## Expected behaviour

`let f a b = A.map (*) a b`
`let arrays x = (fill (constant (Z :. 5)) (constant x)) : (arrays (x + 1))`
`let arrays2 = Prelude.take 10 (arrays 1)`
`streamState arrays2 f (Prelude.take 1 arrays2)`

And the result would be

`Vector (Z :. 10) [3628800, 3628800, 3628800, 3628800, 3628800, 3628800, 3628800, 3628800, 3628800, 3628800] `

The idea is that it will generate a list of arrays, and then feed those arrays to the GPU one at a time,
and the GPU will modify an array in its own memory based off of the incoming arrays. Once the list has
reached an end then it can return the array it was working on in GPU memory.

The idea is that arrays2 might be too big to fit in GPU memory, and the array f returns doesn't need to be sent back and forth across the CPU memory if it is very large.

## Current behaviour

Currently there's nothing like this, and so the only way to handle a problem like this is to drag
large arrays back and forth across slow CPU memory.

## Possible solution (optional)

I think you might be able to get it working by modifying runNWith so that there's an array in GPU
memory that gets saved for the next run.

## Steps to reproduce (for bugs)

 1.
 2.
 3.

## Your environment

- Version used: version 1.2.0.0
- Backend(s) used and version: accelerate-llvm-ptx 1.2.0.0
- GHC version: 8.4.2
- Operating system and version: Debian 8.10
- Link to your project: I haven't uploaded it yet.
- If this is a bug with the GPU backend, include the output of `nvidia-device-query`:

**Comment by tmcdonell** [comment]
Actually, the arrays will only be returned to the CPU once you try and read the array values (i.e. evaluate to normal form; evaluating to WHNF is enough to kick off the computation). They also won't be re-uploaded if they still exist in GPU memory (even if you transferred them back to the CPU for whatever reason).

So I think you just need your function to return the updated state along with the computed value, and then keep feeding the new state in at each iteration. This is a common pattern in simulation codes, for example.

Does that help?

**Comment by skynet1** [comment]
So once I call the run function it calculates the answer, and then sends it back to the CPU, but then when it does it becomes an Array DIM* Elt instead of Acc (Array DIM* Elt) and then I don't have to use the use function because it recognizes the output even though it's not contained in Acc? Does it assign a unique id to the outputs?

**Comment by tmcdonell** [comment]
Calling the `run*` functions do _not_ automatically send the result to the CPU.

The `use` function will recognise when an array already exists on the GPU, and elide the transfer in that case.

So, let's say you have defined your accelerate function similarly to:

```haskell
foo :: (Arrays a, Arrays state)
    => Acc a           -- inputs
    -> Acc state       -- old state
    -> Acc (a, state)  -- result and updated state
```

then the compiled version of this function:

```haskell
foo' :: a -> state -> (a, state)
foo' = runN foo
```

and you can keep calling `foo'`, feeding the updated state and new values each time.

If you compile with debugging enabled, the `-ddump-gc` flag will print a message when an actual data transfer takes place.

**Comment by skynet1** [comment]
Oh. I didn't know Acc could contain more than one thing. Okay. Now I understand.

**Comment by tmcdonell** [comment]
Yes, currently `Acc` can contain nested tuples of arrays (up to 16-tuples, though this is just a matter of writing more type class instances if wider tuples are required).

You can define `Arrays` instances for your own custom data types, although currently this is a little difficult. We are looking into how to make this easy for the next release (e.g. by adding a `deriving Arrays` clause on your data type, or similar).

Anyway, glad I could help. Good luck! ^_^

**Comment by skynet1** [comment]
Wait, there's a problem here. I can't find any documentation anywhere on how to take two arrays in the Acc type, and get an Acc (Array, Array) type.

**Comment by tmcdonell** [comment]
`lift` and `unlift` will do this for you:

https://hackage.haskell.org/package/accelerate-1.2.0.0/docs/Data-Array-Accelerate.html#g:45

As the documentation mentions, sometimes GHC will complain and you'll need to give it a type signature to satisfy it. `TypeApplications` can also help, for example `lift @Acc`

**Comment by skynet1** [comment]
Wait, I can't seem to find any documentation on how to take two Acc Arrays and combine them to get Acc (Array, Array).
